packages/runtime/container-runtime/src/opLifecycle/inbox.ts

```diff
@@ -0,0 +1,592 @@
+	IDocumentSchemaChangeMessage,
+	IDocumentSchemaFeatures,
+} from "../summary/index.js";
+
+import type { DuplicateBatchDetector } from "./duplicateBatchDetector.js";
+import {
+	ensureContentsDeserialized,
+	type BatchStartInfo,
+	type RemoteMessageProcessor,
+} from "./remoteMessageProcessor.js";
+
+interface IInboxEvents extends IEvent {
+	(event: "op", listener: () => void);
+}
+
+/**
+ * ! TODO: remove this interface and put directly on Inbox ctor
+ */
+export interface IInboxParameters {
+	readonly remoteMessageProcessor: RemoteMessageProcessor;
+	readonly duplicateBatchDetector: DuplicateBatchDetector | undefined;
+	readonly pendingStateManager: PendingStateManager;
+	readonly logger: ITelemetryLoggerExt;
+	readonly resetReconnectCount: () => void;
+	readonly ensureNoDataModelChanges: <T>(callback: () => T) => T;
+	readonly updateDocumentDirtyState: (dirty: boolean) => void;
+	readonly hasPendingMessages: () => boolean;
+	readonly channelCollection: ChannelCollection;
+	readonly garbageCollector: IGarbageCollector;
+	readonly blobManager: BlobManager;
+	readonly documentsSchemaController: DocumentsSchemaController;
+	/**
+	 * Id Compressor serializes final state (see getPendingLocalState()). As result, it needs to skip all ops that preceeded that state
+	 * (such ops will be marked by Loader layer as savedOp === true)
+	 * That said, in "delayed" mode it's possible that Id Compressor was never initialized before getPendingLocalState() is called.
+	 * In such case we have to process all ops, including those marked with savedOp === true.
+	 */
+	readonly skipSavedCompressorOps: boolean;
+	readonly getIdCompressor: () => (IIdCompressor & IIdCompressorCore) | undefined;
+	readonly getSessionSchema: () => {
+		[P in keyof IDocumentSchemaFeatures]?: IDocumentSchemaFeatures[P] extends boolean
+			? true
+			: IDocumentSchemaFeatures[P];
+	};
+	readonly closeFn: (error?: ICriticalContainerError) => void;
+	readonly usePendingStateMSN: boolean;
+	readonly initialSequenceNumbers: {
+		snapshotSequenceNumber: number;
+		minimumSequenceNumber: number;
+		lastProcessedSequenceNumber: number;
+	};
+}
+
+/**
+ * ! TODO: move this to more general package and have container runtime interface implement it
+ */
+export interface IOpProcessingProperties {
+	readonly initialSequenceNumber: number;
+	readonly lastProcessedSequenceNumber: number;
+	readonly minimumSequenceNumber: number;
+	readonly lastKnownSeqNumber: number;
+}
+
+export class Inbox extends TypedEventEmitter<IInboxEvents> implements IOpProcessingProperties {
+	constructor(private readonly params: IInboxParameters) {
+		super();
+		this.initialSequenceNumber = params.initialSequenceNumbers.snapshotSequenceNumber;
+		this._actualMinimumSequenceNumber = params.initialSequenceNumbers.minimumSequenceNumber;
+		this._lastProcessedSequenceNumber =
+			params.initialSequenceNumbers.lastProcessedSequenceNumber;
+		this._lastKnownSeqNumber = params.initialSequenceNumbers.lastProcessedSequenceNumber;
```

Comment:
What about checkpointSequenceNumber or whatever we get on connect to Ordering Service?